File: src/Models/Warehouse.cs
namespace StockManagementSystem.Models
{
    public class Warehouse : BaseEntity
    {
        private Warehouse() { }

        public Warehouse(int maximumStockLevel, MaterialType allowedMaterialType)
        {
            MaximumStockLevel = maximumStockLevel;
            AllowedMaterialType = allowedMaterialType;
        }

        public int MaximumStockLevel { get; private set; }

        public MaterialType AllowedMaterialType { get; private set; }

        public int CurrentStockLevel
            => Items.Sum(i => i.Quantity);

        public int FreeStockSpace
            => MaximumStockLevel - CurrentStockLevel;

        public bool HasFreeSpace
            => FreeStockSpace > 0;

        public List<WarehouseItem> Items { get; private set; } = new();

        public List<WarehouseItemLog> ItemsLog { get; private set; } = new();

        public void Import(int productId, int quantity)
        {
            var item = Items.FirstOrDefault(i => i.ProductId == productId);

            if (item == null)
            {
                Items.Add(new WarehouseItem(productId, quantity));
            }
            else
            {
                item.AddQuantity(quantity);
            }
        }

        public void Export(int productId, int quantity)
        {
            var item = Items.First(i => i.ProductId == productId);

            item.RemoveQuantity(quantity);

            if (item.Quantity == 0)
            {
                Items.Remove(item);
            }
        }

        public void AddWarehouseItemLog(
            int productId,
            int previousQuantity,
            int currentQuantity,
            WarehouseOperationType operationType)
        {
            var log = new WarehouseItemLog(
                productId,
                previousQuantity,
                currentQuantity,
                operationType);

            ItemsLog.Add(log);
        }
    }
}

File: src/ClientApp/src/app/warehouse/warehouse.ts
import { MaterialType } from "../product/product";

export interface Warehouse {
    id: number;
    allowedMaterialType: MaterialType;
    maximumStockLevel: number;
    currentStockLevel: number;
    freeStockSpace: number;
    hasFreeSpace: boolean;
    items: WarehouseItem[];
    itemsLog: WarehouseItemLog[];
}

export enum OperationType {
    Import,
    Export
}

export interface WarehouseItem {
    productId: number;
    quantity: number;
}

export interface WarehouseItemLog {
    productId: number;
    previousQuantity: number;
    currentQuantity: number;
    operationType: OperationType;
    createdAtUtc: Date
}

export interface WarehouseRequest {
    productId: number;
    quantity: number;
}